server/src/main/java/org/activityinfo/server/endpoint/rest/DatabaseResource.java

```diff
@@ -147,4 +172,245 @@ public Response createFormFromXForm(@Context UriInfo uri, XForm xForm) {
+
+    private Maybe<User> getUserByEmail(String userEmail) throws NoResultException {
+        User user;
+        try {
+            user = entityManagerProvider.get().createQuery(
+                    "SELECT u FROM User u " +
+                            "WHERE u.email = :email", User.class)
+                    .setParameter("email", userEmail)
+                    .getSingleResult();
+            return Maybe.of(user);
+        } catch (NoResultException e) {
+            return Maybe.notFound();
+        }
+    }
+
+    @POST
+    @Path("/transfer/request")
+    @Consumes(MediaType.APPLICATION_JSON)
+    public Response requestTransfer(@InjectParam AuthenticatedUser executingUser, String json) {
+        RequestTransfer request = RequestTransfer.fromJson(PARSER.parse(json));
+        Database database = getDatabase();
+        User currentOwner = database.getOwner();
+        Maybe<User> proposedOwner = getUserByEmail(request.getProposedOwnerEmail());
+
+        if (executingUser.isAnonymous()) {
+            return Response.status(Response.Status.UNAUTHORIZED).build();
+        }
+        if (executingUser.getUserId() != currentOwner.getId()) {
+            return Response.status(Response.Status.FORBIDDEN).build();
+        }
+        if (database.getTransferToken() != null) {
+            return Response.status(Response.Status.CONFLICT).entity("There is a pending transfer on this database").build();
+        }
+        if (!proposedOwner.isVisible()) {
+            return Response.status(Response.Status.BAD_REQUEST).entity("Proposed owner does not exist").build();
+        }
+        if (currentOwner.equals(proposedOwner.get())) {
+            return Response.status(Response.Status.BAD_REQUEST).entity("User already owns database.").build();
+        }
+
+        startTransaction();
+        String token;
+        try {
+            token = generateTransferToken(database, proposedOwner.get());
+            sendRequestNotifications(currentOwner, proposedOwner.get(), database);
+        } catch (Exception e) {
+            rollbackTransaction();
+            throw new RuntimeException(e);
+        }
+        commitTransaction();
+
+        return Response.ok().entity(token).build();
+    }
+
+    private void sendRequestNotifications(User currentOwner, User proposedOwner, Database database) {
+        mailSender.send(new RequestDatabaseTransferMessage(proposedOwner, currentOwner, database));
+    }
+
+    private String generateTransferToken(Database database, User proposedOwner) {
+        database.setTransferToken(SecureTokenGenerator.generate());
+        database.setTransferUser(proposedOwner);
+        database.setTransferRequestDate(new Date());
+        entityManagerProvider.get().persist(database);
+        return database.getTransferToken();
+    }
+
+    @POST
+    @Path("/transfer/start")
+    @Consumes(MediaType.APPLICATION_JSON)
+    public Response startTransfer(@InjectParam AuthenticatedUser executingUser, String json) {
```

Comment:
Why is this exposed as a REST endpoint? Shouldn't this only be called if the transfeer accepts?

Reply:
Holdover from when I was thinking about an asynchronous approval mechanism. Will change this to a normal method.